### PR TITLE
Add CORS headers for RP-Initiated logout over `POST`

### LINKED ISSUE #32648: RP-Initiated logout using `POST` method fails in cross-origin setup
As part of [the specification](https://openid.net/specs/openid-connect-rpinitiated-1_0.html) we have implemented the `POST` method of RP-Initiated Logout under #12228, and support for Keycloak JS was added under #25348. However, when the RP (client) is on a different origin than the OP (OpenID Provider) this method fails. The reason for this is that the `POST` endpoint does not set the [`Access-Control-Allow-Origin` CORS header ](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSMissingAllowOrigin)required for allowing these sorts of requests.

**Steps to reproduce**

I've [set up a little project](https://github.com/jonkoops/kjs-playground) that I use for testing Keycloak and Keycloak JS in cross-origin scenarios which I will use for convenience of reproducing this issue. Follow the steps outlined in the README and then add `logoutMethod: 'POST'` to the options provided to the `init()` call to Keycloak JS in `src/main.ts`.

**Proposed solution**

When calling the `POST` endpoint the `Access-Control-Allow-Origin` header should be added to the response to allow cross-origin requests. Only valid web origins configured for the client should be allowed at any time. There is already [existing code](https://github.com/keycloak/keycloak/blob/557d7e87b2936dc5b1102ffe5376b283af6bab5f/services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java#L56-L71) that could be re-used to accomplish this for the login status iframe, which need a similar check.

### PR DESCRIPTION
**Pull request by jonkoops**
Closes #32648

### CODE REVIEW

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java [hunk 1]
```diff
@@ -308,6 +308,8 @@ private Response displayLogoutConfirmationScreen(LoginFormsProvider loginForm, A
     @NoCache
     @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
     public Response logout() {
+        Cors.builder().allowAllOrigins().add();
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java:311]: Perhaps we want to allow only the origins configured with the client by looking at the ID Token or the client_id parameter?

Comment (mposolda) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java:311]: +1 , we know the client at this point https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java#L213 , so maybe at this point we can set origins based on that?

**Review by mposolda — COMMENTED** [review]
Added inline comment replying to Pedro's comment. Some more questions:

- In some cases, the logout can involve multiple HTTP requests. For example when "logout confirmation" screen is displayed (but that is probably never the case when logout is triggered by `keycloak.js` adapter as this one is supposed to include `id_token_hint` as well and this means logout confirmation is not displayed) or maybe also when logout contains some front-channel clients (as those display separate iframe for the logout). Do we want to handle this case as well? If yes, there are probably multiple places where we want to set origins.

- Do we want to test this? Or is it too tricky to test and is rather a corner case? For the inspiration, we have some test in `RPInitiatedLogoutTest.logoutWithPostRequest()`, so maybe we can just check if the header is present in the response?

**Comment by jonkoops** [comment]
> In some cases, the logout can involve multiple HTTP requests. For example when "logout confirmation" screen is displayed (but that is probably never the case when logout is triggered by `keycloak.js` adapter as this one is supposed to include `id_token_hint` as well and this means logout confirmation is not displayed) or maybe also when logout contains some front-channel clients (as those display separate iframe for the logout). Do we want to handle this case as well? If yes, there are probably multiple places where we want to set origins.

So the way this works currently in Keycloak JS will follow a redirect if it occurs. If no redirect occurs, it refreshes the page the adapter is embedded in, assuming the session has been ended if the request ends with a status in the 2xx range ([see implementation](https://github.com/keycloak/keycloak/pull/25348/files#diff-eeb93448269ec6cd4a3c7c3b55fd5a77ab95ca0a63ec3d46c13a25b739d29a42R1353-R1360)) . Like you said, since `id_token_hint` is always set,  there should be no need for the additional confirmation.

I am not familiar with these 'front-channel clients', if an iframe would be embedded it would result in a `GET` request. I am not sure if that is a supported use-case as per spec, but if it is we might have to set additional headers to allow embedding the pages in an iframe (depending on our default settings for headers regarding embeds). Perhaps we handle that separately in another issue if we need to?

> Do we want to test this? Or is it too tricky to test and is rather a corner case? For the inspiration, we have some test in `RPInitiatedLogoutTest.logoutWithPostRequest()`, so maybe we can just check if the header is present in the response?

Yes, I want to have some tests for this in place. We do already [have tests in place](https://github.com/keycloak/keycloak/blob/4b95b42590beb695ef0d794b8db6f0517fff7591/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java#L173-L193) to test the `POST` logout with Keycloak JS, but they are currently not being ran in a cross-origin manner in the test suite. I'll add some tests to the `RPInitiatedLogoutTest` to check these headers explicitly.

As for verifying if the client is allowed based on the allowed origins, that is something that can be done perhaps, we already do it for the iframes used for the session status checks and silent login, and I can [re-use that code](https://github.com/keycloak/keycloak/blob/557d7e87b2936dc5b1102ffe5376b283af6bab5f/services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java#L56-L71). What do we do when the `client_id` parameter is not present (as it is optional in the spec)? I can simply omit the headers in that case, but it would also mean the client does not get any sort of error information back, as the request will be blocked by the CORS policy, but perhaps that is fine.

**Review by stianst — CHANGES_REQUESTED** [review]
RP-initiated logout endpoint should not be called directly from JS. It's a redirect endpoint, and should not set any CORS headers.

See the spec:
image

**Comment by jonkoops** [comment]
Closing this PR as @stianst and I have decided to take a different approach. I'll keep the issue around as it is still valid, but we'll be solving this in a different manner.
